internal/operator/uninstall.go

```diff
@@ -0,0 +1,153 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package operator
+
+import (
+	"context"
+	"fmt"
+	"sort"
+	"strings"
+
+	v1 "github.com/operator-framework/api/pkg/operators/v1"
+	"github.com/operator-framework/api/pkg/operators/v1alpha1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/yaml"
+)
+
+type Uninstall struct {
+	config *Configuration
+
+	Package string
+}
+
+func NewUninstall(cfg *Configuration) *Uninstall {
+	return &Uninstall{
+		config: cfg,
+	}
+}
+
+func (u *Uninstall) Run(ctx context.Context) error {
+	subs := v1alpha1.SubscriptionList{}
+	if err := u.config.Client.List(ctx, &subs, client.InNamespace(u.config.Namespace)); err != nil {
+		return fmt.Errorf("list subscriptions: %v", err)
+	}
+
+	var sub *v1alpha1.Subscription
+	for _, s := range subs.Items {
+		if u.Package == s.Spec.Package {
+			sub = &s
+			break
+		}
+	}
+	if sub == nil {
+		return fmt.Errorf("operator package %q not found", u.Package)
+	}
+
+	catsrcKey := types.NamespacedName{
+		Namespace: sub.Spec.CatalogSourceNamespace,
+		Name:      sub.Spec.CatalogSource,
+	}
+	catsrc := &v1alpha1.CatalogSource{}
+	if err := u.config.Client.Get(ctx, catsrcKey, catsrc); err != nil {
+		return fmt.Errorf("get catalog source: %v", err)
+	}
+
+	installPlanKey := types.NamespacedName{
+		Namespace: sub.Status.InstallPlanRef.Namespace,
+		Name:      sub.Status.InstallPlanRef.Name,
+	}
+
+	// Since the install plan is owned by the subscription, we need to
+	// read all of the resource references from the install plan before
+	// deleting the subscription.
+	deleteObjs, err := u.getInstallPlanResources(ctx, installPlanKey)
+	if err != nil {
+		return err
+	}
+
+	// Delete the subscription first, so that no further installs or upgrades
+	// of the operator occur while we're cleaning up.
+	if err := u.config.Client.Delete(ctx, sub); err != nil {
+		return fmt.Errorf("delete subscription %q: %v", sub.Name, err)
+	}
+	fmt.Printf("subscription %q deleted\n", sub.Name)
+
+	// Ensure CustomResourceDefinitions are deleted first, so that the operator
+	// has a chance to handle CRs that have finalizers.
+	sort.SliceStable(deleteObjs, func(i, j int) bool {
+		return deleteObjs[i].GetObjectKind().GroupVersionKind().Kind == "CustomResourceDefinition"
+	})
+	for _, obj := range deleteObjs {
+		err := u.config.Client.Delete(ctx, obj)
+		if err != nil && !apierrors.IsNotFound(err) {
+			return err
+		}
+		if !apierrors.IsNotFound(err) {
+			fmt.Printf("%s %q deleted\n", strings.ToLower(obj.GetObjectKind().GroupVersionKind().Kind), obj.GetName())
+		}
```

Comment:
The case of an unwanted error being returned is already checked above, so the object was either successfully deleted or not found
```suggestion
		}
		fmt.Printf("%s %q deleted\n", strings.ToLower(obj.GetObjectKind().GroupVersionKind().Kind), obj.GetName())
```

Reply:
There are two cases where we don't return `err` above. One is when `err == nil` (the object was deleted), and the other is when `apierrors.IsNotFound(err)` (the object didn't exist).

I only want to print the "deleted" message if the item was _actually_ deleted. Maybe to make this more clear, I can change the if statement to `if err == nil`? WDYT?